README.md

```diff
@@ -1,27 +1,165 @@
+    - [Resolver nodes](#resolver-nodes)
+    - [Config nodes](#config-nodes)
+    - [Pipes](#pipes)
+  - [License](#license)
+
+# Introduction
+
+Funicular is a tool for tight integration of **Clojure full stack** (backend+frontend) apps. The central piece of this integration is the `funicular.edn` schema file.
+
+It's a **tree-like schema** file that acts as a **contract** between backend and frontend parts of the app by defining all the **commands and queries** that the backend system exposes to the frontend, along with a **handler function** and the **input and output [Malli schemas](https://github.com/metosin/malli)** for every particular command and query (`fn + schemas = resolver`).
 
-## Usage
+Resolver nodes can be nested under context nodes that are used to define additional validation rules - you can define an input schema that is going to be validated before going deeper down the tree or an output schema that is going to be validated against the return value.
 
-FIXME: write usage documentation!
+The execution model is based on **[Sieppari](https://github.com/metosin/sieppari) [interceptors](https://quanttype.net/posts/2018-08-03-why-interceptors.html)**. Every resolver node and context node in the schema file is transformed into an interceptor during the compilation phase while preserving the execution order by traversing and compiling the tree-like schema file into a flat list of `:enter` and `:leave` interceptors. These interceptors are then executed sequentially.
 
-Build a deployable jar of this library:
+**Each command and query maps to exactly one resolver function**, before or after which multiple interceptors can be ran.
+
+# `funicular.edn`
+
+An example of a document with 2 queries and 2 commands:
+
+```clojure
+{:api
+ [:api-v1
+  [:content
+   {:interceptors [{:enter log-content-fetch}]
+    :queries
+    {:get-numbers
+     {:input-schema :int
+      :output-schema [:vector :int]
+      :handler (fn [{times :data}] (vec (range times)))}
+     :get-words
+     {:input-schema :app/seed
+      :output-schema :app/generated-words
+      :handler (fn [{times :data}] (mapv make-word (range times)))}}}]
+  [:tracking
+   {:rules logged-in?
+    :commands
+    {:log-visitor
+     {:input-schema :any
+      :output-schema :int
+      :handler (fn [data] (log-site-visit data))}}}]
+  [:util
+   {:queries
+    {:foo
+     {:input-schema :any
+      :output-schema :string
+      :handler (fn [_] :bar)}}
+    :commands
+    {:ping
+     {:input-schema :any
+      :output-schema [:fn #(= % :pong)]
+      :handler (fn [_] :pong)}}}]]}
+```
 
-    $ clojure -A:jar
+To run these queries and commands, you send a request like below:
 
-## Contributions / development
+```clojure
+(ns app.funicular-example
+  (:require [com.verybigthings.funicular.core :as fun]
+            [clojure.edn :as edn]))
 
-Run tests with
+(let [raw-edn (edn/read-string (slurp (io/resource "funicular.edn")))
+      compiled (fun/compile api)]
```

Comment:
Hey. Reading this documentation as inexperienced funicular user and I'm wondering if `api` here is a typo? Shouldn't `raw-edn` be sent to `compile`?

Reply:
Good catch!